src/ucs/arch/generic/cpu.c

```diff
@@ -0,0 +1,94 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/arch/cpu.h>
+#include <ucs/sys/sys.h>
+#include <ucs/sys/string.h>
+
+#define UCS_CPU_CACHE_FILE_FMT     "/sys/devices/system/cpu/cpu%d/cache/index%d/%s"
+#define UCS_CPU_CACHE_LEVEL        "level"
+#define UCS_CPU_CACHE_TYPE         "type"
+#define UCS_CPU_CACHE_SIZE         "size"
+
+/* cache datatypes */
+typedef struct ucs_cpu_cache_info {
+    int  level;
+    char type[32];
+} ucs_cpu_cache_info_t;
+
+
+static ucs_cpu_cache_info_t ucs_cpu_cache[] = {
+    [UCS_CPU_CACHE_L1d] = {.level = 1, .type = "Data"},
+    [UCS_CPU_CACHE_L1i] = {.level = 1, .type = "Instruction"},
+    [UCS_CPU_CACHE_L2]  = {.level = 2, .type = "Unified"},
+    [UCS_CPU_CACHE_L3]  = {.level = 3, .type = "Unified"}
+};
+
+
+ucs_status_t ucs_cpu_get_cache_size(ucs_cpu_cache_type_t type, size_t *value)
+{
+    char type_str[32];  /* Data/Instruction/Unified */
+    char size_str[32];  /* memunits */
+    int i;
+    int cpu;
+    long level;
+    size_t cache_size;
+    ssize_t file_size;
+    ucs_status_t status;
+
+    if (type >= UCS_CPU_CACHE_LAST) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    /* try first CPU-specific algorithm */
+    status = ucs_arch_get_cache_size(type, value);
+    if (status == UCS_OK) {
+        return UCS_OK;
+    }
+
+    cpu = ucs_get_first_cpu();
+
+    for (i = 0;; i++) {
+        file_size = ucs_read_file_str(type_str, sizeof(type_str), 1, UCS_CPU_CACHE_FILE_FMT,
```

Comment:
can we cache the values after 1st use?

Reply:
yep, added, but it made function more complicated